passes/pass4.nim

```diff
@@ -226,21 +242,48 @@ proc colorGraph(gr: var Graph) =
       n.color = next
       inc next
 
+  if conflicts.len > 0:
+    # conflict resolution is needed. First, sort the conflict pairs in
+    # ascending order
+    sort(conflicts, proc (a, b: auto): int = a.a.int - b.a.int)
+    var map: Table[Color, Color]
+
+    # all colors part of a pair must share the same color. Thanks to the
+    # list being sorted by the pairs' first element, only a single iteration
+    # is needed
+    for it in conflicts.items:
+      let
+        a = addr map.mgetOrPut(it.a, it.a)
+        # note: `next` is a known impossible value here. Also, we cannot use
+        # ``mgetOrPut`` for the second color, as that would invalidate `a`
+        b = map.getOrDefault(it.b, next)
+        m = min(a[], b)
+      if m < a[]:
+        a[] = m
+      if m < b:
+        map[it.b] = m
+
+    # now do the actual recoloring:
+    for i, c in gr.conts.pairs:
+      for n in gr.nodes.toOpenArray(c.nodes.a, c.nodes.b).mitems:
+        if n.keep and n.color in map:
+          let newColor = map[n.color]
+          if newColor != n.color:
```

Comment:
If they are the same then setting it would be a no-op, why not always set it and skip the `if`?

Reply:
It's a premature optimization to prevent memory writes when there's no changes, though I don't think it's all that beneficial, nor have I done any measurements for this code. Thank you for the question -- I'll change it.

Reply:
Oh, sorry, I didn't re-read the code before commenting. The guard is necessary to prevent the assertion (which needs to be replaced with proper error reporting through `ReportContext` at some point) from triggering when it shouldn't.

Reply:
> Oh, sorry, I didn't re-read the code before commenting. The guard is necessary to prevent the assertion (which needs to be replaced with proper error reporting through `ReportContext` at some point) from triggering when it shouldn't.

I was thinking that the guard could be moved into the assertion itself and then it could always assign thereafter.